src/lib/handleFileUpload.ts

```diff
@@ -2,9 +2,24 @@ import DAV from './DAV';
 import Entry from './Entry';
 import State from './State';
 import joinPath from './joinPath';
-import { success } from 'melba-toast';
+import { success, error } from 'melba-toast';
 import { t } from 'i18next';
 
+const XHRPutFile = (
+  url: string,
+  file: File,
+  onProgress: (progress: number) => void
+): Promise<XMLHttpRequest> => {
+  return new Promise((resolve, reject) => {
```

Comment:
S: I think this should just replace the `PUT` property in `HTTP`, since it supports that feature. `DAV.upload` could be updated to accept an optional `progressEvent` handler to pass in too. That keeps the API clean and the code separated nicely.

Reply:
Good idea! Will do. Other changes are pretty easy to do, this one is more tricky, though, so might take some time.

BTW, how do you prefer:
* PRs:
  * a) multiple separate commits (like here), or
  * b) all changes in a single commit
* review fixes
  * a) as separate commits (like I just pushed here), or
  * b) squashed into relevant commits

Reply:
> might take some time

Of course! No pressure on timescales from my PoV! :)

For the primary commits, breaking into chunks is perfectly fine, I don't feel particularly strongly about `git` history, meaningful chunks is my preference. For review fixes, it's easier to review if the commits are separate and squashed after the review is complete, but if you squash upfront, that's fine too. Appreciate you checking on these!

Reply:
One more question, if I may:

Currently, `DAV.upload` uses `DAV.#toastOnFailure` to wrap a call to `HTTP.PUT`. `DAV.#toastOnFailure` [expects that a wrapped function will raise `RequestFailure` exception](https://github.com/dom111/webdav-js/blob/master/src/lib/DAV.ts#L89). If new implementation of `HTTP.PUT` will use XMLHttpRequest instead of fetch API, what's better (or less worse):
* in case of failure, for `HTTP.PUT` to raise a manually-constructed `RequestFailure` exception with required fields (is it even possible?), or
* for `DAV.#toastOnFailure` to be extened to inspect a value returned by new implementation of `HTTP.PUT`, or
* for `DAV.upload` to **not** use `DAV.#toastOnFailure`, but instead to inspect a value returned by `HTTP.PUT` and call `error(...)` in case it's needed?

Now that I wrote it, I think last option is the best... Do you agree?

Reply:
Apologies for missing this response, not sure how it got lost in my emails!

I've had a look at this and have a proposed change for you:

https://github.com/Lex-2008/webdav-js/compare/xhr-upload-progress...dom111:webdav-js:xhr-upload-progress-proposal?expand=1

It's option 1 from above and means that the `methodXHR` is pretty much compatible with the other `method`. Hope that make sense!

Reply:
Ah, now I see what you mean. Nice! Actually a similar idea came to my mind, but only after I pushed the change :)

I was also thinking that choice of method (fetch or XHR) could be decided inside the `method` function by the presence of the `onProgress` argument: if it's desired - then use XHR, otherwise - `fetch`... But now that I think about it, XHR can have both upload and download progresses, latter never being used - so maybe it's not a very good idea.

I'll add your change and clean up git history tomorrow. Shall I keep you as the author of latest commit ("Proposed change"), or can I squash it with others and claim authorship? :D To be fair, my typescript knowledge is not good enough to write something like `{ ...parameters, headers: [['Content-Type', file.type]], }` ;)

Reply:
Yeah, I agree having the `HTTP` method choose whether it wants XHR or fetch would be nice, but as you say, it adds some complexity! This is still great and if the need arises could be updated to handle as required!

As for authoring, whatever sits best with you, happy to keep it as a collaboration or have you squash the changes!

Thanks again for your time, greatly appreciated!

Reply:
Sorry for a delay, amended! I also cleaned up git history, so it follows the logic now (first commit introduces methodXHR, second uses it it DAV, third adds GUI, etc). In case I missed something, old commits are available on branch https://github.com/Lex-2008/webdav-js/tree/xhr-upload-progress-bak1

Also note the "nullify xhr.response for "204 No Content" response" commit - probably it should be squashed into first commit, but maybe worth discussing it first. I also don't know if you prefer more verbose `if` approach, like it's currently written, or shorter but more cryptic code with ternary operator, something like `new Response(xhr.status === 204 ? null : xhr.response, {...)`.

Reply:
This looks good as-is! Nice catch on the `Response` needing `null`! Thanks! Just need to look at getting the tests to pass and I think we're good! I'll try and look at it this evening, if time allows! Thanks again for contributing your time and effort!